### ISSUE TITLE
Symbols Reference

### ISSUE THREAD

**Issue by Fireboyd78** [LRFLEW/OpenMC2#8]
Hi, I noticed you haven't used the proper names for classes/functions in most places, so I'm providing this customized .MAP file in hopes that it'll help you figure out some symbol names. It's from Midtown Madness 2 (Beta 2) and we used it in my project [MM2Hook](https://github.com/Fireboyd78/mm2hook) to figure out the names of stuff.

The reason you can use a lot of these names is because both games run on the Angel Game Engine (AGE), and stuff like `phBound` have not changed going forward into the RAGE-era of games!

Link to symbols: https://pastebin.com/h3Ew80Wn

Hope it helps!

**Comment by Fireboyd78** [comment]
I named some of the symbols myself awhile back, here's a partial list of them: https://pastebin.com/uxe5Y2XH

Obviously it's not all of them, but I did try to focus on the important ones such as `datArgParser`, etc.

**EDIT:**
Symbols have been updated, more stuff added

**Comment by LRFLEW** [comment]
Sorry I didn't respond to this quicker.

This all looks interesting. How did you get the original symbol names for MM2 in the first place? Did you manually match them in the other pastebin you posted, or did you use something like FLARE to do it procedurally? If you can send me the executable that the original MAP file refers to, I could use that to try and identify more symbols.

I'm not 100% sure if I want to use the same names as the original in this project. Renaming things can help with the organization of the project, and avoid falling into any pitfalls that the original might have (I can already see quite a bit of bad code in the game). However, this will at the very least be useful as reference. I'll have to look over this at some point, but there's a lot of symbols to manually go over to update the names in the project.

**Comment by Fireboyd78** [comment]
Of course, I totally understand. The symbols I linked to you were from the second beta of Midtown Madness 2. However, I manually matched all symbols to the XP-patch for Midtown Madness 2. That was quite an experience!

Midtown Madness 2:
If you use IDA Pro, apply [this script](https://pastebin.com/rWNincyJ) to a database, and it'll name everything for you. Just ignore errors about "reserved prefixes".

Also provided is the custom map file for the retail XP patch of MM2 [here](https://pastebin.com/h5wQuNqC). It's easy to find on either MM2X or MMArchive.

Midnight Club 2:
[Here's](https://pastebin.com/uRHhBxjH) the script for MC2 symbols. Not sure why I didn't think to provide that, lol.

**Comment by InfamousKnight** [comment]
Has this project been abandoned? I would love to tas midnight club 2 with libtas pretty soon..

**Comment by LRFLEW** [comment]
This project hasn't been abandoned. The problem is that there's basically one dev who actually knows how to do the reverse engineering work for this, and he (ie. me) just started a Masters program. I'm still around, and if you ping me on the Midnight Club speedrunning Discord, I will (usually) respond. I've been meaning to do a sort-of rebase of this project, given the new symbols provided here and some antivirus complaints about the DLL, but I just haven't had the time to work on that.

I'm interested in seeing what you have in mind for TASing, though. I recently started participating in the SM64 TAS Competition(s), so would be interested in following (and perhaps contributing to) such a TAS. I don't have any experience with libtas, though, so I have no idea how well it would work for this. I wouldn't recommend using OpenMC2 for the TAS as of right now, though. The project isn't at a point where it will make things particularly easier, and I have reason to believe I may have introduced a crash or two. In the (far) future, I plan on making some changes that might make a native TAS system viable, but that's a long ways off (and lower priority than other changes, such as improved net code).

**Comment by InfamousKnight** [comment]
What do you mean by native TAS system? Libtas should be able to handle this project once wine isn’t required anymore. As in libtas doesn’t support wine yet. So once native Linux support is made, this should be tasable. I was thinking about making entertainment speed trade offs like destroying motorcycles with a car and hitting every pedestrian while winning the race. Or even getting in front of those chases where you follow the opponent to the race and only press left or right to balance as you go there with it.

**Comment by Krutonium** [comment]
I heard about libtas at GDQ, they said never again will they do anything different. I will admit it's making me curious.

**Comment by LRFLEW** [comment]
@Fireboyd78 I'm finally getting around to making a big update to this project, and I'm trying to include some of the symbols you found in it. (At least, I'm using it as a guide. For example, the logging functions have names like `Displayf`, which I don't like, but I am renaming the function to `mc2_log_display` as a compromise).

Looking over some of the symbols, I noticed something that would explain something I found but didn't understand. Can you do me a favor and check if the MM2 address `0x5773A0` (`??_U@YAPAXI@Z`) matches the MC2 address `0x5772A0`. You can also check if the MM2 address `0x5773C0` (`??_V@YAXPAX@Z`) matches the MC2 address `0x577350`, but I assume that it will if the first one does. If they do match up, I can use it to get rid of some Undefined Behavior that I had thought I had to rely on.

**Comment by Fireboyd78** [comment]
@LRFLEW Nice catch! Yes, those functions do seem to be the `new[]` and `delete[]` operators, respectively. You can safely name the addresses using those symbol names.

**Comment by LRFLEW** [comment]
After I made the comment before, I took a second look at the functions, and now I'm not sure about my original assessment.

The issue comes down to `memMemoryAllocator::Allocate` (and `Free`), and how the second argument works. I can see from the original signature that the value is a `bool`, though the signature and a quick scan of the function doesn't immediately reveal its purpose. My working hypothesis is that the `bool` argument indicates if the allocation is an array allocation.

This would make sense, as the function at `0x577320` uses `false` as `operator delete` and the function at `0x577350` uses `true` as `operator delete[]`. However, this doesn't work for the allocators, as both the function at `0x5772E0` (`operator new` with the debug parameters) and the function at `0x5772A0` (the one I identified as `operator new[]`) pass in `false` for the parameter. This made me think that maybe the function at `0x5772A0` is actually also `operator new` (but without the debug parameters). However, it doesn't make sense for there to be an `operator delete[]` but no `operator new[]` (at least to me). It's possible that using `false` in the function at `0x5772A0` was a mistake. However, if it is `operator new[]`, I would expect it to be called about as often as `operator delete[]`, but the number of references is 26 and 136 respectively, which seems like an excessively large difference.

Do you have any thoughts on this? I was thinking that if I hooked `operator new` and `operator new[]` correctly, then I would get rid of some technical undefined behavior, but right now I'm not confident enough to make that change.

**Comment by Fireboyd78** [comment]
I'll double check for you, but I felt pretty confident naming those functions as the new/delete array operators. Don't think too hard on this stuff, if you make a mistake it's easy to correct later on. Their memory allocator system was vastly improved since MM2 so it's harder to find stuff. And yes, the bool is for array data.
